core/tls-websocket-server/src/ws_server.rs

```diff
@@ -18,71 +18,427 @@
+		&self,
+		poll: &mut mio::Poll,
+		event: &Event,
+		signal_receiver: &mut Receiver<ServerSignal>,
+	) -> WebSocketResult<bool> {
+		let signal = signal_receiver.try_recv()?;
+
+		let initiate_shut_down = match signal {
+			ServerSignal::ShutDown => true,
+		};
+
+		signal_receiver.reregister(
+			poll,
+			event.token(),
+			mio::Ready::readable(),
+			mio::PollOpt::level(),
+		)?;
+
+		Ok(initiate_shut_down)
+	}
+
+	fn register_server_signal_sender(&self, sender: Sender<ServerSignal>) -> WebSocketResult<()> {
+		let mut sender_lock =
+			self.signal_sender.lock().map_err(|_| WebSocketError::LockPoisoning)?;
+		*sender_lock = Some(sender);
+		Ok(())
 	}
 }
 
-impl WebSocketServer for TungsteniteWsServer {
-	type Connection = TungsteniteWsConnection;
+impl<Handler, ConfigProvider> WebSocketServer for TungsteniteWsServer<Handler, ConfigProvider>
+where
+	ConfigProvider: ProvideServerConfig,
+	Handler: WebSocketMessageHandler,
+{
+	type Connection = TungsteniteWsConnection<Handler>;
 
-	fn run<Handler>(&self, handler: Arc<Handler>) -> WebSocketResult<()>
-	where
-		Handler: WebSocketHandler<Connection = Self::Connection>,
-	{
+	fn run(&self) -> WebSocketResult<()> {
 		debug!("Running tungstenite web socket server on {}", self.ws_address);
 
 		let socket_addr: SocketAddr =
 			self.ws_address.parse().map_err(WebSocketError::InvalidWsAddress)?;
 
-		let config = make_config(self.cert_path.as_str(), self.private_key_path.as_str())?;
+		let config = self.config_provider.get_config()?;
+
+		let (server_signal_sender, mut shutdown_receiver) = channel::<ServerSignal>();
+		self.register_server_signal_sender(server_signal_sender)?;
+
+		let tcp_listener = TcpListener::bind(&socket_addr).map_err(WebSocketError::TcpBindError)?;
+		let mut poll = Poll::new()?;
+		poll.register(
+			&tcp_listener,
+			NEW_CONNECTIONS_LISTENER,
+			mio::Ready::readable(),
+			mio::PollOpt::level(),
+		)?;
+
+		poll.register(
+			&shutdown_receiver,
+			SERVER_SIGNAL_TOKEN,
+			mio::Ready::readable(),
+			mio::PollOpt::level(),
+		)?;
+
+		let mut events = mio::Events::with_capacity(1024);
+
+		// Run the event loop.
+		'outer_event_loop: loop {
+			poll.poll(&mut events, None)?;
+
+			for event in events.iter() {
+				match event.token() {
+					NEW_CONNECTIONS_LISTENER => {
+						debug!("Received new connection event");
+						if let Err(e) =
+							self.accept_connection(&mut poll, &tcp_listener, config.clone())
+						{
+							error!("Failed to accept new web-socket connection: {:?}", e);
+						}
+					},
+					SERVER_SIGNAL_TOKEN => {
+						debug!("Received server signal event");
+						if self.handle_server_signal(&mut poll, &event, &mut shutdown_receiver)? {
+							break 'outer_event_loop
+						}
+					},
+					_ => {
+						debug!("Connection (token {:?}) activity event", event.token());
+						if let Err(e) = self.connection_event(&mut poll, &event) {
+							error!("Failed to process connection event: {:?}", e);
+						}
+					},
+				}
+			}
+		}
```

Comment:
This is the main event loop for `mio`. We have 3 types of events: 
1. New TCP connections to be accepted
2. Existing connections have activity
3. Server signal, such as shut-down request